asset normalization

soulstompp opened the pull request:
I am currently still working on getting postgres up to date with SQLite but wanted to get this request open for commenting as I worked through that. Any input appreciated.

xemwebe commented:
Thanks for the update, I see you have made some substantial progress. So far it looks quite good. Here are just a few remarks:

1. I have seen that you have changed the signature of `insert_asset` to return an Asset struct instead of just the new asset id. What is the reasoning behind that? Usually, you should know what the asset is you try to insert. I admit that it simplifies dealing with the case where an asset of that name already exists in the database.
2. You have named the id fields of the new `Stock` and `Currency` structs `asset_id`, while the id of `Asset` is just id (as in the other cases). This is a bit confusing. I would suggest to name id field `id` in all cases (in consistency with the primary key of the respective database table), but add the struct name to any id reference the id in another struct (like `asset_id` in transaction), which translate to foreign keys in database tables. What do you think?
3. I think the `rounding_digets` field of pg currency tables misses an `INT NOT NULL` specifier.

soulstompp commented:
On Sun, Jan 30, 2022, 2:34 PM Mark Beinker ***@***.***> wrote:

> Thanks for the update, I see you have made some substantial progress. So
> far it looks quite good. Here are just a few remarks:
>
>    1. I have seen that you have changed the signature of insert_asset to
>    return an Asset struct instead of just the new asset id. What is the
>    reasoning behind that? Usually, you should know what the asset is you try
>    to insert. I admit that it simplifies dealing with the case where an asset
>    of that name already exists in the database.
>
> I was concerned with the fact that we were going to have to update the
resources object id as well. However, thinking this through further, I
should have just updated Asset::set_id() to set the resources id to its id.
I think it would be best to change that signature back and have the
consumer call set_id for their object.

>
>    1. You have named the id fields of the new Stock and Currency structs
>    asset_id, while the id of Asset is just id (as in the other cases).
>    This is a bit confusing. I would suggest to name id field id in all
>    cases (in consistency with the primary key of the respective database
>    table), but add the struct name to any id reference the id in another
>    struct (like asset_id in transaction), which translate to foreign keys
>    in database tables. What do you think?
>
> Agreed, I like that added consistency.

>
>    1. I think the rounding_digets field of pg currency tables misses an INT
>    NOT NULL specifier.
>
> Oops, I'll get that one before I start trying to get pg tests passing.

>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/xemwebe/finql/pull/1#issuecomment-1025248116>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAXYWZAOTMS2DPRN6FORE3UYW4I3ANCNFSM5NDZEXBA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

soulstompp commented:
I've made changes to reflect your 3 recommendations above and am going to try diving into the pg side of things now.

soulstompp commented:
Not sure if it was the best approach, but I amended my partial commit from before and replaced it. This is ready for closer review now, since I have the pg portion of this compiling properly and updated the sqlx query macro json data. Please let me know if you see anything I have missed.

xemwebe commented:
Great, I will have a look at it later this day.

xemwebe commented on finql-data/src/asset.rs:
```diff
@@ -2,35 +2,76 @@ use super::{DataError, DataItem};
+pub enum Resource {
+    Currency(Currency),
+    Stock(Stock),
 }
```
I think AssetClass would a better name for this enum than Resource. I would also expect this enum to grow over time, i.e. including "FixedIncome", "Funds", etc.

xemwebe commented on finql-data/src/asset.rs:
```diff
@@ -53,6 +94,16 @@ impl DataItem for Asset {
+                        s.id = Some(id);
+                    }
+                }
+
```
Storing the id at two different places, e.g. at the level of the Asset and at the level of the AssetClass/Resource is calling for trouble, since you have to ensure that they are always consistent. Why to you think its required to store the id on AssetClass/Resource level, too?

xemwebe commented on finql-data/src/currency.rs:
```diff
@@ -46,14 +46,41 @@ impl de::Error for CurrencyError {
+        }
+    }
+}
+
```
Reviewing my own old code I would now prefer a more compact implementation of this method, e.g. in the lines of
```rust
if len(code)!= 3 {
   return Err(...);
}
let iso_code = code.iter()
    .map(|c| c.to_ascii_uppercase() )
    .filter(|c| c.is_ascii_alphabetic()).collect();
Ok(Self { iso_code } )
```
However, this would not guarantee that the final result has exactly 3 chars and `collect()` probably does not work for fixed size arrays. Therefore, its probably a bit more complex than the above suggestion...

xemwebe commented on finql-sqlite/src/asset_handler.rs:
```diff
@@ -3,34 +3,52 @@
+                        params![id, &s.isin, &s.wkn])?;
+                },
+            }
+
```
This is problematic. Image what happens when you try to insert a currency that already exists in the database. The insert above for the asset will pass without problems, but the second query trying to insert the currency will fill. As result, you will end up with a new asset in the database with an id for which no currency exists. 
In general, I think it's no good design to have the requirement that there must be an asset in the database for every stock or currency or vice versa. I'm thinking of a solution where we could avoid storing the asset in the database.

xemwebe commented on finql-sqlite/src/transaction_handler.rs:
```diff
@@ -136,92 +134,134 @@ impl TransactionHandler for SqliteDB {
+                        cash_currency: Currency::new(
+                            row.get(4)?,
+                            CurrencyISOCode::from_str(&cic).expect("expected good code from the database"),
+                            row.get(6)?,
```
Please no expect or unwrap anywhere in the core library! Library code that is used in other programs should never panic. It's okay to unwrap if the value has been checked previously to be okay, but not otherwise. The database entry could be corrupted, therefore I would return an error instead of using `expect`. 
In tests and examples `unwrap` or `expect` is no problem.

xemwebe commented on finql-sqlite/src/transaction_handler.rs:
```diff
@@ -136,92 +134,134 @@ impl TransactionHandler for SqliteDB {
+                    &transaction.asset,
+                    &transaction.cash_amount,
+                    &transaction.cash_currency.id.expect("cash currency asset id must be in db"),
+                    &transaction.cash_date,
```
see below, no use of `expect` here.

xemwebe commented on finql-sqlite/src/transaction_handler.rs:
```diff
@@ -136,92 +134,134 @@ impl TransactionHandler for SqliteDB {
                     &transaction.cash_amount,
-                    &transaction.cash_currency,
+                    &transaction.cash_currency.id.expect("cash currency asset id must be in db"),
                     &transaction.cash_date,
```
see below, no use of `expect`

xemwebe reviewed (changes requested):
That's quite a large commit, to large to be approved just by looking at it. Also, I'm not happy with the new role of the Asset struct which is now just a wrapper around the actual assets of different asset classes (see also the comments in the review). Also, the Asset struct no longer holds any meaningful data. Not all assets need a unique name (including currencies and stock options, for example). The mere function of the Asset is now to provide some unique interface to different types of asset classes, which in rust would be better represented by a trait.

I would prefer a solution, where there is no need to store this Asset Wrapper in the database at all. Binding a certain asset to a transaction could be solved in rust by using an id that consists of a simple `enum` and the id of the asset in its asset class specific data base table. However, we need also a solution to represent this on database level without losing referential integrity. So far I have no good solution to this problem, I have to think of it.

xemwebe commented:
After thinking a bit about this PR, I would suggest to change the approach

1. Removing the asset struct and renaming Resource to Asset, i.e. Asset will become an enum
2. In the database, I would follow your approach of storing assets in tables for each asset class with one extra asset table for providing unique ids for all assets. Storing and deleting assets however need to be embedded in a transaction, enabling a rollback in case something goes wrong. This should guarantee consistency of the database state.

I will merge your branch as a new branch into the main repository and implement the above approach to see how far we get with this.

soulstompp commented:
Sounds like a good plan. I should have time later in the evening if I can
be of assistance.

A domingo, 3/04/2022, 12:40, Mark Beinker ***@***.***>
escreveu:

> After thinking a bit about this PR, I would suggest to change the approach
>
>    1. Removing the asset struct and renaming Resource to Asset, i.e.
>    Asset will become an enum
>    2. In the database, I would follow your approach of storing assets in
>    tables for each asset class with one extra asset table for providing unique
>    ids for all assets. Storing and deleting assets however need to be embedded
>    in a transaction, enabling a rollback in case something goes wrong. This
>    should guarantee consistency of the database state.
>
> I will merge your branch as a new branch into the main repository and
> implement the above approach to see how far we get with this.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/xemwebe/finql/pull/1#issuecomment-1086846384>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAXYW7BIGJIBPMD4AOXUSDVDF7TVANCNFSM5NDZEXBA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

xemwebe commented:
Since it is quite a burden to support both databases, Sqlite and
PostgreSQL, and this also limits efficient database usage to some
extent, I consider dropping support for Sqlite. 

Do you have any preferences regarding Sqlite or PostgreSQL support?

Am Sonntag, dem 03.04.2022 um 04:59 -0700 schrieb Kenny Flegal:
> Sounds like a good plan. I should have time later in the evening if I
> can
> be of assistance.
> 
> A domingo, 3/04/2022, 12:40, Mark Beinker ***@***.***>
> escreveu:
> 
> > After thinking a bit about this PR, I would suggest to change the
> approach
> >
> > 1. Removing the asset struct and renaming Resource to Asset, i.e.
> > Asset will become an enum
> > 2. In the database, I would follow your approach of storing assets
> in
> > tables for each asset class with one extra asset table for
> providing unique
> > ids for all assets. Storing and deleting assets however need to be
> embedded
> > in a transaction, enabling a rollback in case something goes wrong.
> This
> > should guarantee consistency of the database state.
> >
> > I will merge your branch as a new branch into the main repository
> and
> > implement the above approach to see how far we get with this.
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <https://github.com/xemwebe/finql/pull/1#issuecomment-1086846384>,
> or
> > unsubscribe
> > <https://github.com/notifications/unsubscribe-
> auth/AAAXYW7BIGJIBPMD4AOXUSDVDF7TVANCNFSM5NDZEXBA>
> > .
> > You are receiving this because you authored the thread.Message ID:
> > ***@***.***>
> >
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.Message ID:
> ***@***.***>

soulstompp commented:
I think Postgres would be a much better fit for the project even if it does
require a bit more effort. In particular it would allow for us to use table
inheritance for tables like the assets parent table, which handled the
complications of this approach elegantly.

A domingo, 3/04/2022, 14:18, Mark Beinker ***@***.***>
escreveu:

> Since it is quite a burden to support both databases, Sqlite and
> PostgreSQL, and this also limits efficient database usage to some
> extent, I consider dropping support for Sqlite.
>
> Do you have any preferences regarding Sqlite or PostgreSQL support?
>
> Am Sonntag, dem 03.04.2022 um 04:59 -0700 schrieb Kenny Flegal:
> > Sounds like a good plan. I should have time later in the evening if I
> > can
> > be of assistance.
> >
> > A domingo, 3/04/2022, 12:40, Mark Beinker ***@***.***>
> > escreveu:
> >
> > > After thinking a bit about this PR, I would suggest to change the
> > approach
> > >
> > > 1. Removing the asset struct and renaming Resource to Asset, i.e.
> > > Asset will become an enum
> > > 2. In the database, I would follow your approach of storing assets
> > in
> > > tables for each asset class with one extra asset table for
> > providing unique
> > > ids for all assets. Storing and deleting assets however need to be
> > embedded
> > > in a transaction, enabling a rollback in case something goes wrong.
> > This
> > > should guarantee consistency of the database state.
> > >
> > > I will merge your branch as a new branch into the main repository
> > and
> > > implement the above approach to see how far we get with this.
> > >
> > > —
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/xemwebe/finql/pull/1#issuecomment-1086846384>,
> > or
> > > unsubscribe
> > > <https://github.com/notifications/unsubscribe-
> > auth/AAAXYW7BIGJIBPMD4AOXUSDVDF7TVANCNFSM5NDZEXBA>
> > > .
> > > You are receiving this because you authored the thread.Message ID:
> > > ***@***.***>
> > >
> > —
> > Reply to this email directly, view it on GitHub, or unsubscribe.
> > You are receiving this because you commented.Message ID:
> > ***@***.***>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/xemwebe/finql/pull/1#issuecomment-1086867324>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAXYW2UWIU7KY7DQ2UNOTTVDGLDBANCNFSM5NDZEXBA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

xemwebe commented:
I have done not most of the changes in branch currency-normalization, but some errors are still remaining

xemwebe commented:
I have merged your branch which some modifications and updates of mine.

soulstompp commented:
That's great news! I will look through the changes tonight.

A quarta, 6/04/2022, 17:37, Mark Beinker ***@***.***>
escreveu:

> I have merged your branch which some modifications and updates of mine.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/xemwebe/finql/pull/1#issuecomment-1090475019>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAXYW6VAQATLXEPCWTVCDDVDW4T5ANCNFSM5NDZEXBA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

xemwebe commented:
Though the approach does work now, it becomes apparent that something has to be done to about handling the ID field. I consider enforcing (at least by default) to store assets in the database at creation time to ensure they have always the id field set and hide also ID handling away from the library user, or even make the ID field private.

soulstompp commented:
I think saving it immediately by defuapt may be the best approach. I have
been meaning to play with doing this via table inheritance but haven't had
a chance yet. However, if it does help this would be a very postgres
specific solution.

On Thu, Apr 7, 2022, 7:36 AM Mark Beinker ***@***.***> wrote:

> Though the approach does work now, it becomes apparent that something has
> to be done to about handling the ID field. I consider enforcing (at least
> by default) to store assets in the database at creation time to ensure they
> have always the id field set and hide also ID handling away from the
> library user, or even make the ID field private.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/xemwebe/finql/pull/1#issuecomment-1091133822>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAXYWYUERTNF346MM6R3EDVDZ7AHANCNFSM5NDZEXBA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
